### PR TITLE
[ELY-1062] - add way to defer listener registration

### PR DESCRIPTION
**Pull request by baranowb**
https://issues.jboss.org/browse/ELY-1062
Enabler: https://issues.jboss.org/browse/JBEAP-9684

### CODE REVIEW

File: src/main/java/org/wildfly/security/_private/ElytronMessages.java [hunk 1]
```diff
@@ -590,8 +590,8 @@
     @Message(id = 1145, value = "Security realm [%s] must implement [%s]")
     IllegalArgumentException realmCacheUnexpectedType(SecurityRealm realm, Class<? extends CacheableSecurityRealm> expectedType);
 
-    @Message(id = 1146, value = "Ldap realm failed to register notification listener")
-    IllegalStateException ldapRealmFailedRegisterListener(@Cause  Throwable cause);
+    //@Message(id = 1146, value = "Ldap realm failed to register notification listener")
+    //IllegalStateException ldapRealmFailedRegisterListener(@Cause  Throwable cause);
```
Comment (honza889) [inline src/main/java/org/wildfly/security/_private/ElytronMessages.java]: I think you can remove the removed message completely and put the new message on its place - no need to keep it commented out here.

File: src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java [hunk 2]
```diff
@@ -130,6 +135,7 @@
         this.credentialLoaders = credentialLoaders;
         this.credentialPersisters = credentialPersisters;
         this.evidenceVerifiers = evidenceVerifiers;
+        this.name = name;
```
Comment (honza889) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: I am not sure if this is not unsystematic - something like keeping name of realm/keystore/credentialstore etc. would be useful on a lot of other places too - I think we should add it to all of not at all...

File: src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java [hunk 3]
```diff
@@ -147,17 +153,30 @@ public void registerIdentityChangeListener(Consumer<Principal> listener) {
         DirContext dirContext = null;
         try {
             dirContext = obtainContext();
-            EventContext eventContext = (EventContext) dirContext.lookup("");
-            eventContext.addNamingListener("", EventContext.SUBTREE_SCOPE, new ServerNotificationListener(listener));
+            innerIdentityChangeListener(dirContext,listener);
         } catch (Exception cause) {
-            throw log.ldapRealmFailedRegisterListener(cause);
+            //throw log.ldapRealmFailedRegisterListener(cause);
```
Comment (honza889) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: Please, remove completely, no reason to keep it in comments.

File: src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java [hunk 4]
```diff
@@ -147,17 +153,30 @@ public void registerIdentityChangeListener(Consumer<Principal> listener) {
...
+                log.debug("Listener registration failure: ", cause);
+            }
         } finally {
             if (dirContext != null) {
                 closeContext(dirContext);
             }
         }
     }
 
+    private void innerIdentityChangeListener(final DirContext dirContext, final Consumer<Principal> listener) throws NamingException {
```
Comment (honza889) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: Name of method should say what the method do - like `registerIdentityChangeListener`

File: src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java [hunk 5]
```diff
@@ -181,10 +200,26 @@ private ModifiableRealmIdentity getRealmIdentity(final Principal principal, fina
     }
 
     private DirContext obtainContext() throws RealmUnavailableException {
-        try {
-            return dirContextSupplier.get();
-        } catch (NamingException e) {
-            throw log.ldapRealmFailedToObtainContext(e);
+        if (this.listenersPendingRegistration.size() == 0) {
```
Comment (honza889) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: Much more importent: if `listenersPendingRegistration` requires synchronization, all its usages should be in synchronized block.

File: src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java [hunk 6]
```diff
@@ -181,10 +200,26 @@ private ModifiableRealmIdentity getRealmIdentity(final Principal principal, fina
...
+                return dirContextSupplier.get();
+            } catch (NamingException e) {
+                throw log.ldapRealmFailedToObtainContext(e);
+            }
+        } else {
+            try {
+                DirContext ctx = dirContextSupplier.get();
+                synchronized (this.listenersPendingRegistration) {
+                    //we got ctx, this means connection is up
+                    for(Consumer<Principal> listener:this.listenersPendingRegistration){
```
Comment (honza889) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: Minor, but when you will be already in updating, please add spaces around `:` to keep codestyle.

File: src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java (hunk 5 above)
Comment (honza889) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: Moreover, the big condition is useless here - the `else` part is sufficient (as the condition does not bring any optimization after adding correct synchronization).

File: src/main/java/org/wildfly/security/_private/ElytronMessages.java (hunk 1 above)
Comment (baranowb) [inline src/main/java/org/wildfly/security/_private/ElytronMessages.java]: Ok, had no idea, usually WFLY has those commented out.

File: src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java (hunk 2 above)
Comment (baranowb) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: True, but this is out of scope of this JIRA/PR

File: src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java [hunk 7]
```diff
@@ -147,17 +153,29 @@ public void registerIdentityChangeListener(Consumer<Principal> listener) {
...
+                log.debug("Listener registration failure: ", cause);
+            }
         } finally {
             if (dirContext != null) {
                 closeContext(dirContext);
             }
         }
     }
 
+    private void innerRegisterIdentityChangeListener(final DirContext dirContext, final Consumer<Principal> listener) throws NamingException {
```
Comment (baranowb) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: heh, what I meant is that this PR/JIRA does not warrant change in other resources. Change that this PR has is there in order to be able to identify realm which is not "fully up".
WRT to method name, there is "registerIdentityChangeListener", second one is private, if you want I can prefix it with _ if this is more suitable style.

File: src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java (hunk 2 above)
Comment (honza889) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: Ok, as reminded @darranl : 
> We could have an ELY issue as well then but I think it was a deliberate decision that Elytron components don't need to be named.
> Maybe a higher level discussion in the future about how we bridge that gap.

So adding the name need to be removed from PR. Sorry.

File: src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java (hunk 7 above)
Comment (honza889) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: The problem is the prefixing itself - it is bad practise called Hungarien notation. You should not put such information into name when you are able to do it on type system level using `private` keyword and similar.
http://stackoverflow.com/questions/111933/why-shouldnt-i-use-hungarian-notation

Comment (baranowb) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: well, I just was oblivious to fact that both methods have different set of params. ech...

Comment (honza889) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: Sorry, I didnt notice that the same method name is already used for different method - in such case is "inner" prefix ok (if there is no better name, which would explain difference between them)

File: src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java [hunk 8]
```diff
@@ -182,7 +197,15 @@ private ModifiableRealmIdentity getRealmIdentity(final Principal principal, fina
 
     private DirContext obtainContext() throws RealmUnavailableException {
         try {
-            return dirContextSupplier.get();
+            synchronized (this.listenersPendingRegistration) {
+                DirContext ctx = dirContextSupplier.get();
```
Comment (honza889) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: As context obtaining is network operation often (which can be time consuming everytime), it would be better to have it outside of the synchronized section.

File: src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java (hunk 7 above)
Comment (baranowb) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: No need, it can have the same name and just be private, since method signature will be different due to params.

File: src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java [hunk 9]
```diff
@@ -147,17 +150,29 @@ public void registerIdentityChangeListener(Consumer<Principal> listener) {
         DirContext dirContext = null;
         try {
             dirContext = obtainContext();
-            EventContext eventContext = (EventContext) dirContext.lookup("");
-            eventContext.addNamingListener("", EventContext.SUBTREE_SCOPE, new ServerNotificationListener(listener));
+            registerIdentityChangeListener(dirContext,listener);
```
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/auth/realm/ldap/LdapSecurityRealm.java]: Won't there be a potential race like this:

Thread 1 | Thread 2
------|------
registerIdentityChangeListener fails|
-|connection succeeds
-|listeners pending registration added (empty)
listener added to pending registration list|

And then the listener is never called?

**Comment by honza889** [comment]
@baranowb Need to fix race condition (as demonstrated by David above) + rebase

**Comment by honza889** [comment]
Looks resolved... returning +1 back

**Comment by dmlloyd** [comment]
Needs a rebase.

**Comment by darranl** [comment]
If we can get this rebased I can probably get it merged and into the tag I am creating today.
